File: frontend/src/app/game/maze.ts
import { MazeData } from './services';

export interface Cell {
  x: number;
  y: number;
  walls: {
    top: boolean;
    right: boolean;
    bottom: boolean;
    left: boolean;
  };
  visited: boolean;
}

export class Maze {
  cells: Cell[][] = [];
  width: number;
  height: number;

  constructor(width: number, height: number, serverData?: MazeData) {
    this.width = width;
    this.height = height;

    if (serverData) {
      this.loadFromServer(serverData);
    } else {
      this.generate();
    }
  }

  /**
   * Load maze from server data
   */
  private loadFromServer(data: MazeData): void {
    this.cells = [];
    for (let y = 0; y < data.height; y++) {
      const row: Cell[] = [];
      for (let x = 0; x < data.width; x++) {
        const serverCell = data.cells[y][x];
        row.push({
          x: serverCell.x,
          y: serverCell.y,
          walls: {
            top: serverCell.top,
            right: serverCell.right,
            bottom: serverCell.bottom,
            left: serverCell.left,
          },
          visited: true, // Server maze is already generated
        });
      }
      this.cells.push(row);
    }
  }

  /**
   * Create maze from server data (static factory)
   */
  static fromServer(data: MazeData): Maze {
    return new Maze(data.width, data.height, data);
  }

  private generate(): void {
    // Initialize grid with all walls
    this.cells = [];
    for (let y = 0; y < this.height; y++) {
      const row: Cell[] = [];
      for (let x = 0; x < this.width; x++) {
        row.push({
          x,
          y,
          walls: { top: true, right: true, bottom: true, left: true },
          visited: false,
        });
      }
      this.cells.push(row);
    }

    // Recursive backtracking algorithm
    const stack: Cell[] = [];
    const startCell = this.cells[0][0];
    startCell.visited = true;
    stack.push(startCell);

    while (stack.length > 0) {
      const current = stack[stack.length - 1];
      const neighbors = this.getUnvisitedNeighbors(current);

      if (neighbors.length === 0) {
        stack.pop();
      } else {
        const next = neighbors[Math.floor(Math.random() * neighbors.length)];
        this.removeWall(current, next);
        next.visited = true;
        stack.push(next);
      }
    }
  }

  private getUnvisitedNeighbors(cell: Cell): Cell[] {
    const neighbors: Cell[] = [];
    const { x, y } = cell;

    if (y > 0 && !this.cells[y - 1][x].visited) {
      neighbors.push(this.cells[y - 1][x]);
    }
    if (x < this.width - 1 && !this.cells[y][x + 1].visited) {
      neighbors.push(this.cells[y][x + 1]);
    }
    if (y < this.height - 1 && !this.cells[y + 1][x].visited) {
      neighbors.push(this.cells[y + 1][x]);
    }
    if (x > 0 && !this.cells[y][x - 1].visited) {
      neighbors.push(this.cells[y][x - 1]);
    }

    return neighbors;
  }

  private removeWall(current: Cell, next: Cell): void {
    const dx = next.x - current.x;
    const dy = next.y - current.y;

    if (dx === 1) {
      current.walls.right = false;
      next.walls.left = false;
    } else if (dx === -1) {
      current.walls.left = false;
      next.walls.right = false;
    } else if (dy === 1) {
      current.walls.bottom = false;
      next.walls.top = false;
    } else if (dy === -1) {
      current.walls.top = false;
      next.walls.bottom = false;
    }
  }

  canMove(fromX: number, fromY: number, toX: number, toY: number): boolean {
    if (toX < 0 || toX >= this.width || toY < 0 || toY >= this.height) {
      return false;
    }

    const current = this.cells[fromY][fromX];
    const dx = toX - fromX;
    const dy = toY - fromY;

    if (dx === 1) return !current.walls.right;
    if (dx === -1) return !current.walls.left;
    if (dy === 1) return !current.walls.bottom;
    if (dy === -1) return !current.walls.top;

    return false;
  }
}

File: websocket-server/internal/game/maze.go
package game

import (
	"fmt"
	"math/rand"
)

// Cell represents a single cell in the maze
type Cell struct {
	X       int  `json:"x"`
	Y       int  `json:"y"`
	Top     bool `json:"top"`
	Right   bool `json:"right"`
	Bottom  bool `json:"bottom"`
	Left    bool `json:"left"`
	Visited bool `json:"-"` // Don't send to client
}

// Maze represents the game maze
type Maze struct {
	Width  int      `json:"width"`
	Height int      `json:"height"`
	Cells  [][]Cell `json:"cells"`
}

// NewMaze generates a new maze using recursive backtracking
func NewMaze(width, height int) *Maze {
	// Note: rand is auto-seeded in Go 1.20+

	// Initialize grid with all walls
	cells := make([][]Cell, height)
	for y := 0; y < height; y++ {
		cells[y] = make([]Cell, width)
		for x := 0; x < width; x++ {
			cells[y][x] = Cell{
				X:       x,
				Y:       y,
				Top:     true,
				Right:   true,
				Bottom:  true,
				Left:    true,
				Visited: false,
			}
		}
	}

	maze := &Maze{
		Width:  width,
		Height: height,
		Cells:  cells,
	}

	// Generate maze using recursive backtracking
	maze.generate()

	// Debug: print cell (0,0) walls
	c := maze.Cells[0][0]
	fmt.Printf("Cell (0,0) walls - Top:%v Right:%v Bottom:%v Left:%v\n", c.Top, c.Right, c.Bottom, c.Left)

	return maze
}

func (m *Maze) generate() {
	stack := []struct{ x, y int }{{0, 0}}
	m.Cells[0][0].Visited = true

	iterations := 0
	for len(stack) > 0 {
		current := stack[len(stack)-1]
		neighbors := m.getUnvisitedNeighbors(current.x, current.y)

		if len(neighbors) == 0 {
			stack = stack[:len(stack)-1] // Pop
		} else {
			// Pick random neighbor
			idx := rand.Intn(len(neighbors))
			next := neighbors[idx]

			if iterations < 5 {
				fmt.Printf("Iteration %d: current=(%d,%d) next=(%d,%d)\n",
					iterations, current.x, current.y, next.x, next.y)
			}

			m.removeWall(current.x, current.y, next.x, next.y)
			m.Cells[next.y][next.x].Visited = true
			stack = append(stack, next)
		}
		iterations++
	}
	fmt.Printf("Maze generated in %d iterations\n", iterations)
}

func (m *Maze) getUnvisitedNeighbors(x, y int) []struct{ x, y int } {
	var neighbors []struct{ x, y int }

	// Up
	if y > 0 && !m.Cells[y-1][x].Visited {
		neighbors = append(neighbors, struct{ x, y int }{x, y - 1})
	}
	// Right
	if x < m.Width-1 && !m.Cells[y][x+1].Visited {
		neighbors = append(neighbors, struct{ x, y int }{x + 1, y})
	}
	// Down
	if y < m.Height-1 && !m.Cells[y+1][x].Visited {
		neighbors = append(neighbors, struct{ x, y int }{x, y + 1})
	}
	// Left
	if x > 0 && !m.Cells[y][x-1].Visited {
		neighbors = append(neighbors, struct{ x, y int }{x - 1, y})
	}

	return neighbors
}

func (m *Maze) removeWall(x1, y1, x2, y2 int) {
	dx := x2 - x1
	dy := y2 - y1

	if dx == 1 {
		m.Cells[y1][x1].Right = false
		m.Cells[y2][x2].Left = false
	} else if dx == -1 {
		m.Cells[y1][x1].Left = false
		m.Cells[y2][x2].Right = false
	} else if dy == 1 {
		m.Cells[y1][x1].Bottom = false
		m.Cells[y2][x2].Top = false
	} else if dy == -1 {
		m.Cells[y1][x1].Top = false
		m.Cells[y2][x2].Bottom = false
	}
}

// CanMove checks if movement from one cell to another is valid
func (m *Maze) CanMove(fromX, fromY, toX, toY int) bool {
	if toX < 0 || toX >= m.Width || toY < 0 || toY >= m.Height {
		return false
	}

	cell := m.Cells[fromY][fromX]
	dx := toX - fromX
	dy := toY - fromY

	if dx == 1 {
		return !cell.Right
	}
	if dx == -1 {
		return !cell.Left
	}
	if dy == 1 {
		return !cell.Bottom
	}
	if dy == -1 {
		return !cell.Top
	}

	return false
}
